### ISSUE TITLE
sendTransaction rpc 

### ISSUE THREAD

**Issue by sakshamapp** [amjuarez/bytecoin#171]
https://wiki.bytecoin.org/wiki/Send_transaction_-_Bytecoin_RPC_Wallet_API
According to this ,if i pass any address in params that will send transaction , there is no security .

**Comment by pmitchev** [comment]
Users are not supposed to use walletd directly.

**Comment by System96** [comment]
No, you need to have access to that wallet either through a .wallet file or its private keys, if we had no security I would have already been rich by now :P

**Comment by indigencoin** [comment]
How do you spend the coins from a specific address in a walletcontainer file? I'm trying to send coins with a perfectly formatted json but I keep getting this error: -32600 Invalid Request

**Comment by indigencoin** [comment]
How do you spend the coins from a specific address in a walletcontainer file? I'm trying to send coins with a perfectly formatted json but I keep getting this error: -32600 Invalid Request

**Comment by pmitchev** [comment]
Share the json you use. You are probably misusing a parameter

**Comment by indigencoin** [comment]
{
  "jsonrpc": "2.0",
  "id": "",
  "method": "sendTransaction",
  "params": {
    "addresses": [ "indgn7KayssFJVj4bhzv6LbdG8RF5h3FJJ6utXGqpwAQijBg6GRkCxELHfpZKdNFr8fEF9cPizJGea3YvGsVzppm9NKT8X7vNBT38" ],
    "transfers": [
      {
        "amount": 70000000,
        "address": indgn7KaiUdG7oMfRTEmDh5JcqZf7wMJrg185MnzNF6Y5gBwCjnZbDQLHfpZKdNFr8fEF9cPizJGea3YvGsVzppm9NKT8X7eWrr3E"
      }
    ],
    "fee": 10000,
    "unlockTime": 1,
    "anonymity": 7,
    "extra": "",
    "paymentId": "10",
    "changeAddress": "indgn7KayssFJVj4bhzv6LbdG8RF5h3FJJ6utXGqpwAQijBg6GRkCxELHfpZKdNFr8fEF9cPizJGea3YvGsVzppm9NKT8X7vNBT38"
  },
  "result": { "transactionHash": null }
}

**Comment by pmitchev** [comment]
https://github.com/amjuarez/bytecoin/blob/d56f1c51664d9868403480867e8e68df49993e1e/src/PaymentGate/WalletService.cpp#L836

It looks like it's "**sourceAddresses**", not "**addresses**"

**Comment by indigencoin** [comment]
Same error.

{
  "jsonrpc": "2.0",
  "id": "",
  "method": "sendTransaction",
  "params": {
    "sourceAddresses": [ "indgn7KayssFJVj4bhzv6LbdG8RF5h3FJJ6utXGqpwAQijBg6GRkCxELHfpZKdNFr8fEF9cPizJGea3YvGsVzppm9NKT8X7vNBT38" ],
    "transfers": [
      {
        "amount": 70000000,
        "address": "indgn7KaiUdG7oMfRTEmDh5JcqZf7wMJrg185MnzNF6Y5gBwCjnZbDQLHfpZKdNFr8fEF9cPizJGea3YvGsVzppm9NKT8X7eWrr3E"
      }
    ],
    "fee": 10000,
    "unlockTime": 1,
    "anonymity": 7,
    "extra": "",
    "paymentId": "19",
    "changeAddress": "indgn7KayssFJVj4bhzv6LbdG8RF5h3FJJ6utXGqpwAQijBg6GRkCxELHfpZKdNFr8fEF9cPizJGea3YvGsVzppm9NKT8X7vNBT38"
  },
  "result": { "transactionHash": null }
}

**Comment by indigencoin** [comment]
Which file exactly holds the details of CryptoNote::TransactionParameters?

CryptoNote::TransactionParameters sendParams;
sendParams.sourceAddresses = request.sourceAddresses;

I'm trying to figure out what is contained within "sourceAddresses" if its an array of a strings or an array of objects similar to "transfers".

**Comment by indigencoin** [comment]
Ok I got the source. And it's an array of string. So there must something else. 

struct WalletRpcOrder {
  std::string address;
  uint64_t amount;

  void serialize(CryptoNote::ISerializer& serializer);
};

struct SendTransaction {
  struct Request {
    std::vector<std::string> sourceAddresses;
    std::vector<WalletRpcOrder> transfers;
    std::string changeAddress;
    uint64_t fee = 0;
    uint32_t anonymity = DEFAULT_ANONYMITY_LEVEL;
    std::string extra;
    std::string paymentId;
    uint64_t unlockTime = 0;

    void serialize(CryptoNote::ISerializer& serializer);
  };

  struct Response {
    std::string transactionHash;

    void serialize(CryptoNote::ISerializer& serializer);
  };
};

**Comment by indigencoin** [comment]
Is there any use for the secretSpendKey when sending coins from a specific address?

**Comment by indigencoin** [comment]
Found a fix. It was paymentId parameter that is wrong. It should be payment_id.

But it's kind of strange because both addresses and sourceAddresses are working and valid so you can use either of the two. However, if you use sourceAddresses, it requires changeAddress parameter, while if you use addresses only, it is not required.

**Comment by lanaiscool** [comment]
FInally after 2 days of searching found this post 


bytecoin/src/PaymentGate/WalletService.cpp

Line 836 in d56f1c5
 sendParams.sourceAddresses = request.sourceAddresses; 

It looks like it's "sourceAddresses", not "addresses"

this should be changed in the api documentation

**Comment by pmitchev** [comment]
I was in mistake. The documentation is right. Deleted my comment so others do not get confused
